File: tifs_2014.py
import numpy as np
from PIL import Image
import jpegio as jio
from scipy.fftpack import dct, idct
import os

def bdct(block):
    """Apply 2D blockwise DCT."""
    return dct(dct(block.T, norm='ortho').T, norm='ortho')

def jpeg_rec_gray(jpeg_struct):
    """Reconstruct grayscale image from JPEG DCT coefficients and quant tables."""
    quant_table = jpeg_struct.quant_tables[0]
    coeffs = jpeg_struct.coef_arrays[0]
    h, w = coeffs.shape
    rec = np.zeros_like(coeffs, dtype=np.float32)

    for i in range(0, h, 8):
        for j in range(0, w, 8):
            block = coeffs[i:i+8, j:j+8]
            dequant_block = block * quant_table
            rec[i:i+8, j:j+8] = idct(idct(dequant_block.T, norm='ortho').T, norm='ortho')

    rec = np.clip(np.round(rec), 0, 255).astype(np.uint8)
    return rec

def TIFS_2014(image_paths):
    final_feat = []
    error_images = []
    dct_error_images = []
    trunc_block = []
    round_block = []

    for img_path in image_paths:
        jpeg_img = jio.read(img_path)
        I = np.array(Image.open(img_path).convert('L'))  # grayscale
        rec = jpeg_rec_gray(jpeg_img)

        Q = jpeg_img.quant_tables[0]
        Q_rep = np.tile(Q, (I.shape[0]//8, I.shape[1]//8))

        R = I.astype(float) - rec.astype(float)
        M = np.round(bdct(R) / Q_rep)

        err = R.copy()
        dct_err = bdct(err)

        zero_8 = np.zeros((8, 8))

        r_error = []
        r_error_dc = []
        r_error_ac = []
        t_error = []
        t_error_dc = []
        t_error_ac = []

        trun = 0

        for i in range(0, M.shape[0], 8):
            for j in range(0, M.shape[1], 8):
                M_n = M[i:i+8, j:j+8]

                if np.count_nonzero(M_n == 0) != 64:
                    R_n = R[i:i+8, j:j+8]
                    W_n = (M_n * Q).flatten()

                    if np.max(R_n) <= 0.5 and np.min(R_n) >= -0.5:
                        r_error.append(R_n)
                        r_error_dc.append(W_n[0])
                        r_error_ac.extend(W_n[1:])
                    else:
                        trun = 1
                        t_error.append(R_n)
                        t_error_dc.append(W_n[0])
                        t_error_ac.extend(W_n[1:])

        if trun == 0:
            trunc_block.append(0)
            round_block.append(1)
        else:
            trunc_block.append(1)
            round_block.append(0)

        # Convert lists to arrays for feature calculation
        r_error = np.concatenate([block.flatten() for block in r_error]) if r_error else np.array([0])
        t_error = np.concatenate([block.flatten() for block in t_error]) if t_error else np.array([0])
        r_error_dc = np.array(r_error_dc) if r_error_dc else np.array([0])
        r_error_ac = np.array(r_error_ac) if r_error_ac else np.array([0])
        t_error_dc = np.array(t_error_dc) if t_error_dc else np.array([0])
        t_error_ac = np.array(t_error_ac) if t_error_ac else np.array([0])

        # Feature vector
        feature_vec = [
            np.mean(np.abs(r_error)), np.var(np.abs(r_error)),
            np.mean(np.abs(t_error)), np.var(np.abs(t_error)),
            np.mean(np.abs(r_error_dc)), np.var(np.abs(r_error_dc)),
            np.mean(np.abs(r_error_ac)), np.var(np.abs(r_error_ac)),
            np.mean(np.abs(t_error_dc)), np.var(np.abs(t_error_dc)),
            np.mean(np.abs(t_error_ac)), np.var(np.abs(t_error_ac)),
            len(r_error_dc) / (len(r_error_dc) + len(t_error_dc) + 1e-8)
        ]

        final_feat.append(feature_vec)
        error_images.append(err[np.newaxis, np.newaxis, :, :])
        dct_error_images.append(dct_err[np.newaxis, np.newaxis, :, :])

    return (
        np.array(trunc_block),
        np.array(round_block),
        np.concatenate(dct_error_images, axis=0),
        np.concatenate(error_images, axis=0),
        np.array(final_feat)
    )

File: TIFS_2014.m
function [trunc_block, round_block, dct_error_images, error_images, final_feat] = TIFS_2014(image_path)
    final_feat = [];
    error_images = cell(length(image_path), 1);  % Use cells instead of arrays
    dct_error_images = cell(length(image_path), 1);  % Use cells instead of arrays
    trunc_block = [];
    round_block = [];
    
    for img = 1:length(image_path)  % Changed parfor to for to troubleshoot
        try
            trun = 0;
            img_path = image_path{img,1};
            
            % Print image path and size
            I = imread(img_path);
            fprintf('Processing image %d: %s, Size: %dx%d\n', img, img_path, size(I,1), size(I,2));
            
            jpeg_img = jpeg_read(img_path);
            rec = jpeg_rec_gray(jpeg_img);
            
            % Ensure both images have the same size
            [h, w] = size(I);
            h_new = floor(h/8) * 8;
            w_new = floor(w/8) * 8;
            
            I = I(1:h_new, 1:w_new);
            rec = rec(1:h_new, 1:w_new);
            
            Q = jpeg_img.quant_tables{1,1};
            Q_rep = repmat(Q, h_new/8, w_new/8);
            
            R = double(I) - rec;
            M = int64(bdct(R)./Q_rep);
            err = R;
            dct_err = bdct(err);
            
            % Store in cell arrays instead of concatenating
            err_reshaped = reshape(err, [1, 1, size(err,1), size(err,2)]);
            dct_err_reshaped = reshape(dct_err, [1, 1, size(dct_err,1), size(dct_err,2)]);
            
            error_images{img} = err_reshaped;
            dct_error_images{img} = dct_err_reshaped;
            
            zero_8 = zeros(8,8);
            
            r_error = [];
            r_error_dc = [];
            r_error_ac = [];
            t_error = [];
            t_error_dc = [];
            t_error_ac = [];

            for i = 1:8:size(M,1)
                for j = 1:8:size(M,2)
                    if i+7 <= size(M,1) && j+7 <= size(M,2)  % Boundary check
                        M_n = M(i:i+7, j:j+7);
                        if (nnz(M_n == zero_8) ~= 64)
                            R_n = R(i:i+7, j:j+7);
                            W_n = double(M_n).*Q;
                            W_n = reshape(W_n, 1, 64);
                            
                            if(max(max(R_n)) <= 0.5 && min(min(R_n)) >= -0.5)
                                trun = 0;
                                r_error = [r_error, R_n];
                                r_error_dc = [r_error_dc, W_n(1)];
                                r_error_ac = [r_error_ac, W_n(2:end)];
                            else
                                trun = 1;
                                t_error = [t_error, R_n];
                                t_error_dc = [t_error_dc, W_n(1)];
                                t_error_ac = [t_error_ac, W_n(2:end)];
                            end
                        end
                    end
                end
            end
      
            if(trun == 0)
                trunc_block = [trunc_block, 0];
                round_block = [round_block, 1];
            else
                trunc_block = [trunc_block, 1];
                round_block = [round_block, 0];
            end
            
            % Handle empty arrays for feature calculation
            if isempty(r_error)
                r_error_mean = 0;
                r_error_var = 0;
            else
                r_error_mean = mean(abs(r_error(:)));
                r_error_var = var(abs(r_error(:)));
            end
            
            if isempty(t_error)
                t_error_mean = 0;
                t_error_var = 0;
            else
                t_error_mean = mean(abs(t_error(:)));
                t_error_var = var(abs(t_error(:)));
            end
            
            % Similarly handle other feature calculations
            if isempty(r_error_dc)
                r_error_dc_mean = 0;
                r_error_dc_var = 0;
            else
                r_error_dc_mean = mean(abs(r_error_dc));
                r_error_dc_var = var(abs(r_error_dc));
            end
            
            if isempty(r_error_ac)
                r_error_ac_mean = 0;
                r_error_ac_var = 0;
            else
                r_error_ac_mean = mean(abs(r_error_ac));
                r_error_ac_var = var(abs(r_error_ac));
            end
            
            if isempty(t_error_dc)
                t_error_dc_mean = 0;
                t_error_dc_var = 0;
            else
                t_error_dc_mean = mean(abs(t_error_dc));
                t_error_dc_var = var(abs(t_error_dc));
            end
            
            if isempty(t_error_ac)
                t_error_ac_mean = 0;
                t_error_ac_var = 0;
            else
                t_error_ac_mean = mean(abs(t_error_ac));
                t_error_ac_var = var(abs(t_error_ac));
            end
            
            % Calculate ratio with checks to avoid division by zero
            if isempty(r_error_dc) && isempty(t_error_dc)
                ratio = 0;
            elseif isempty(t_error_dc)
                ratio = 1;
            else
                ratio = length(r_error_dc) / (length(r_error_dc) + length(t_error_dc));
            end
            
            feature_vec = [r_error_mean, r_error_var, ...
                          t_error_mean, t_error_var, ...
                          r_error_dc_mean, r_error_dc_var, ...
                          r_error_ac_mean, r_error_ac_var, ...
                          t_error_dc_mean, t_error_dc_var, ...
                          t_error_ac_mean, t_error_ac_var, ...
                          ratio];
                          
            final_feat = [final_feat; feature_vec];
            
        catch e
            fprintf('Error processing image %d: %s\n', img, e.message);
            fprintf('Stack: %s\n', e.getReport);
            
            % Add placeholder data for this failed image
            trunc_block = [trunc_block, 0];
            round_block = [round_block, 0];
            error_images{img} = [];
            dct_error_images{img} = [];
            final_feat = [final_feat; zeros(1, 13)];
        end
    end
    
    % Convert cell arrays back to matrices if needed for your application
    % This would need custom handling based on how you use these outputs
end